src/scanner.c

```diff
@@ -0,0 +1,173 @@
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <tree_sitter/parser.h>
+
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+enum TokenType {
+    HERE_STRING_BODY,
+};
+
+typedef struct {
+    uint32_t len;
+    uint32_t cap;
+    char *data;
+} String;
```

Comment:
The comment above should be placed here.

> consider write comment that points out the invariant of String: cap == strlen(data) + 1.

This is unnecessary as there are comment in `string_new` that points out there are a null character `\0`.

> change all uint32_t to size_t

`uint32_t` is always 32 bits, `size_t` is 32 bits on 32-bit platform, and 64 bits on 64-bit platform.
(It's not precise, see [cppreference - arithmetic types](https://en.cppreference.com/w/c/language/arithmetic_types))

`uint32_t` could overflow for very large source code (4GB+).

It may be crazy that someone has such a large source code file.
But it's possible.
A more important reason is that parser, is the basic component of a general programming language tool. We should not suppose what people use this language for and what this language is designed for. All we can do is let parser stable, correct, and reliable.

You need to change all variable that indicate the size of a sequence of bytes to `size_t`.

Reply:
with that rationale - wouldn't uint64_t be better

Reply:
See [Stackoverflow - What is size_t in C?](https://stackoverflow.com/questions/2550774/what-is-size-t-in-c)

TLDR. `size_t` is a type guaranteed to hold any array index. It scales very well.
On 32-bit platform, `uint64_t` is unnecessary large, and even the pointer size (or the address space size) is 32-bit..

Reply:
BTW, all these subtle details is why I didn't use C at the first time.

Reply:
I don't particularly think such a check is really important...is it really rational to account for handling heredoc strings more than 4gb :shrug:

Reply:
The language allows it. Even if the language grammar is very crazy and out of control (it's not here), the parser should do it.

The thinking of details and the pursuit of high-quality code makes a different programmer.

What would Linus do if he's here?